.pre-commit-config.yaml

```diff
@@ -0,0 +1,35 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v3.2.0
+    hooks:
+    -   id: trailing-whitespace
+    -   id: end-of-file-fixer
+    -   id: check-yaml
+    -   id: check-added-large-files
+
+  - repo: local
+    hooks:
+      - id: clang-format-fix
+        name: clang-format-fix
+        entry: clang-format-18
+        files: ^(src/|include/)
+        types_or: [c++, c]
+        language: system
+        args: ['-i']
+        stages: [manual]
+
+  - repo: https://github.com/pocc/pre-commit-hooks
+    rev: v1.3.5
+    hooks:
+      - id: clang-format
+        additional_dependencies: ['clang-format==18.1.8']
+      - id: clang-tidy
+        stages: [manual]
+        additional_dependencies: ['clang-tidy==18.1.8']
```

Comment:
This doesn't work, we will need to pass in project structure information to clang-tidy:

```
clang-tidy...............................................................Failed
- hook id: clang-tidy
- exit code: 1

/home/bradwu/Desktop/exemplar/examples/identity_direct_usage.cpp:3:10: error: 'beman/exemplar/identity.hpp' file not found [clang-diagnostic-error]
    3 | #include <beman/exemplar/identity.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "examples/identity_direct_usage.cpp"
No compilation database found in /home/bradwu/Desktop/exemplar/examples or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
1 error generated.
Error while processing /home/bradwu/Desktop/exemplar/examples/identity_direct_usage.cpp.
Found compiler error(s).
/home/bradwu/Desktop/exemplar/src/beman/exemplar/identity.t.cpp:3:10: error: 'beman/exemplar/identity.hpp' file not found [clang-diagnostic-error]
    3 | #include <beman/exemplar/identity.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error while trying to load a compilation database:
Could not auto-detect compilation database for file "src/beman/exemplar/identity.t.cpp"
No compilation database found in /home/bradwu/Desktop/exemplar/src/beman/exemplar or any parent directory
fixed-compilation-database: Error while opening fixed database: No such file or directory
json-compilation-database: Error while opening JSON database: No such file or directory
Running without flags.
1 error generated.
Error while processing /home/bradwu/Desktop/exemplar/src/beman/exemplar/identity.t.cpp.
Found compiler error(s).
```

We will need to provide compilation database and "include" directive (both for header and gtest). 
This means enable `CMAKE_EXPORT_COMPILE_COMMANDS` in cmake and provide `-Iinclude/` (as well gtest equivalent, which depends on the build directory).
A better approach maybe integrating clang-tidy into cmake build infrastructure as a build target and invoking that target in pre-commit.

Needs help here @bretbrownjr @camio @steve-downey @neatudarius

Reply:
Very tempted to exclude this change from this pr.

Reply:
It's perfectly reasonable to not include clang-tidy as part of this PR. I'm also fine with it being part of a CMake build.

Reply:
Cool. I will look into clang-tidy later.

Reply:
resolve to delegate to future pr